component/loki/process/stages/stage.go

```diff
@@ -0,0 +1,221 @@
+}
+
+func toStage(p Processor) Stage {
+	return &stageProcessor{
+		Processor: p,
+		inspector: newInspector(os.Stderr, runtime.GOOS == "windows"),
+	}
+}
+
+// New creates a new stage for the given type and configuration.
+func New(logger log.Logger, jobName *string, cfg StageConfig, registerer prometheus.Registerer) (Stage, error) {
+	var (
+		s   Stage
+		err error
+	)
+	switch {
+	// case StageTypeDocker:
+	// 	s, err = NewDocker(logger, registerer)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeCRI:
+	// 	s, err = NewCRI(logger, registerer)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	case cfg.JSONConfig != nil:
+		s, err = newJSONStage(logger, cfg.JSONConfig)
+		if err != nil {
+			return nil, err
+		}
+	// case StageTypeLogfmt:
+	// 	s, err = newLogfmtStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeRegex:
+	// 	s, err = newRegexStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeMetric:
+	// 	s, err = newMetricStage(logger, cfg, registerer)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	case cfg.LabelsConfig != nil:
+		s, err = newLabelStage(logger, cfg.LabelsConfig)
+		if err != nil {
+			return nil, err
+		}
+	// case StageTypeLabelDrop:
+	// 	s, err = newLabelDropStage(cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeTimestamp:
+	// 	s, err = newTimestampStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeOutput:
+	// 	s, err = newOutputStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeMatch:
+	// 	s, err = newMatcherStage(logger, jobName, cfg, registerer)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeTemplate:
+	// 	s, err = newTemplateStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeTenant:
+	// 	s, err = newTenantStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeReplace:
+	// 	s, err = newReplaceStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeDrop:
+	// 	s, err = newDropStage(logger, cfg, registerer)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeLimit:
+	// 	s, err = newLimitStage(logger, cfg, registerer)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeMultiline:
+	// 	s, err = newMultilineStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypePack:
+	// 	s, err = newPackStage(logger, cfg, registerer)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeLabelAllow:
+	// 	s, err = newLabelAllowStage(cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// case StageTypeStaticLabels:
+	// 	s, err = newStaticLabelsStage(logger, cfg)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	default:
+		return nil, errors.Errorf("Unknown stage type")
```

Comment:
nit: include the stage type provided in the error message

Reply:
I think that right now, we'd never reach this case. We don't switch on stage names anymore, but rather on which of the fields is non-nil.

Either we were able to successfully decode into exactly one of the defined config types, or we weren't which would fail during the Unmarshaling. This may change if we find a good way to handle `stage.json`, so I'm leaving the default case as-is for now.